src/cargo/ops/cargo_package.rs

```diff
@@ -82,48 +83,33 @@ struct GitVcsInfo {
     sha1: String,
 }
 
+/// Packages a single package in a workspace, returning the resulting tar file.
+///
+/// Ignores `opts.list`.
 pub fn package_one(
     ws: &Workspace<'_>,
     pkg: &Package,
     opts: &PackageOpts<'_>,
-) -> CargoResult<Option<FileLock>> {
-    let gctx = ws.gctx();
-    let mut src = PathSource::new(pkg.root(), pkg.package_id().source_id(), gctx);
-    src.update()?;
+) -> CargoResult<FileLock> {
+    let ar_files = build_ar_list(ws, pkg, opts)?;
+    let tarball = create_package(ws, pkg, ar_files)?;
 
-    if opts.check_metadata {
-        check_metadata(pkg, gctx)?;
-    }
-
-    if !pkg.manifest().exclude().is_empty() && !pkg.manifest().include().is_empty() {
-        gctx.shell().warn(
-            "both package.include and package.exclude are specified; \
-             the exclude list will be ignored",
-        )?;
+    if opts.verify {
+        run_verify(ws, pkg, &tarball, opts)?;
     }
-    let src_files = src.list_files(pkg)?;
-
-    // Check (git) repository state, getting the current commit hash if not
-    // dirty.
-    let vcs_info = if !opts.allow_dirty {
-        // This will error if a dirty repo is found.
-        check_repo_state(pkg, &src_files, gctx)?
-    } else {
-        None
-    };
 
-    let ar_files = build_ar_list(ws, pkg, src_files, vcs_info)?;
+    Ok(tarball)
+}
 
+// Builds a tarball and places it in the output directory.
+fn create_package(
+    ws: &Workspace<'_>,
+    pkg: &Package,
+    ar_files: Vec<ArchiveFile>,
+) -> CargoResult<FileLock> {
+    let gctx = ws.gctx();
     let filecount = ar_files.len();
 
-    if opts.list {
```

Comment:
imo `package-one` should still do this.  `cargo publish` might not need it but from an API perspective.

Alternatively, we assert if its set

Reply:
My main motivation here was simplify the return value, because the old behavior was to return `None` when `--list` was set and `Some` otherwise. That was a bit awkward given that the only current caller hard-codes `opts.list` to `false`.

If it's ok with you, I'll go with assert + a comment explaining the situation.